rotate.py

```diff
@@ -33,54 +36,95 @@
 # to get the ID, then we have to get column keys from that.
 # curl -H "Accept: application/vnd.github.inertia-preview+json" -H "Authorization: token OAUTH_TOKEN_HERE" https://api.github.com/projects/4032465/columns
 
-# Here's a horrible global variable
-first = False
-# TODO: make a generator that returns the next working day
-# then we can skip this global crap
-
-
 @click.command()
 @click.option('--firstdate', prompt='The first weekday to start with',
               help='The first weekday to start with. Format: YYYY-MM-DD')
 def make_cards(firstdate):
-    """Takes the first date and makes a set of cards for each given column."""
-    click.echo("OK, making cards starting with {0}".format(firstdate))
-    global first
-    if not first:
-        first = firstdate
-    for column in COLUMNS:
-        create_card(get_two(first), column)
+    """
+    Create rotation cards starting at a specific date.
 
+    rotations: 2 is the default.
+    The rotations is the number of times one individual will do rotations.
+    """
+    click.echo("OK, making cards starting with {0}".format(firstdate))
+    days_cards = create_rotation_days(firstdate, rotations=2)
+    for card in days_cards:
+        create_card(card)
 
-def get_next_workday(today, set_global=True):
-    tomorrow = today + datetime.timedelta(days=1)
-    if tomorrow.isoweekday() in set((6, 7)):
-        tomorrow += datetime.timedelta(days=8 - tomorrow.isoweekday())
-    if set_global:
-        global first
-        first = get_next_workday(tomorrow, False).strftime('%Y-%m-%d')
-    return tomorrow
 
+def create_rotation_days(firstdate, rotations):
+    """
+    Create a list of rotation days to create as card.
 
-def get_two(firstdate):
-    """Return the 2 days we want to create assignments for, as a tuple."""
+    The format is:
+    [(person_name, column_id, day_1, day_2), …]
+    """
+    days_card = []
+    start_date = get_start_date(firstdate)
+    working_days = build_workdays(start_date, rotations)
+    while working_days:
+        for person in COLUMNS:
+            # First day
+            first_day = working_days[0]
+            if first_day.weekday() == COLUMNS[person]['avoid']:
```

Comment:
I'm not sure I fully understand how the `avoid` stuff works. 

I made local modifications to try to understand:

So, for 2021-01-26, with karl set to avoid day `1` (Tuesdays?) (the 26th is a Tuesday)
![Screen Shot 2021-01-20 at 11 46 57 AM](https://user-images.githubusercontent.com/67283/105214155-3e253a00-5b15-11eb-9345-c339444deb4c.png)

That seems to work as expected. Dennis ends up with Tuesday (and Fri), and Karl gets Weds and Thurs. Nice.

But if I pick the 25th, which is a Monday:

![Screen Shot 2021-01-20 at 11 48 19 AM](https://user-images.githubusercontent.com/67283/105214321-788ed700-5b15-11eb-8614-d3cda820f21f.png)

Karl ends up with Monday and Tuesday. Is that expected?

Also, because why not, I set everyone to avoid `1` and it explodes:

![Screen Shot 2021-01-20 at 11 49 30 AM](https://user-images.githubusercontent.com/67283/105214417-93f9e200-5b15-11eb-83ff-2cf9123f1515.png)

(that seems like a use-case to not care about, until team members needs to avoid the same days).

Reply:
hehe we need tests… :D

Reply:
For example, I can't do diagnosis on Tuesdays. Too many meetings.

1. List of people
2. Each person has to do 2 days of diagnosis in one rotation
3. We can set the number of rotations where each individual will do diagnosis. (default 2 rotations == 4 days)
4. Take the first person, take the first day
    1. Check if it's a week-end, find the next monday
    2. Check if it's an authorized day (not in avoid list)
    3. Pick the first possible day from the pile.

The pile of business days is composed at  the beginning, and then we go through this pile of days, always taking the first *authorized* day available. :) 

But I haven't tested all scenarios so there are bugs and this requires tests. :) So that would be something to add to this PR.

Reply:
If people are fine with being assigned **one** day, not two consecutive days, the implementation can be very naive.

```python
import datetime

# 'blocking' is a list of weekdays to block, 0=monday.
PEOPLE = {
  "Dennis": {"blocking": [], "assignments": []},
  "Karl": {"blocking": [1, 3], "assignments": []},
  "Ksenia": {"blocking": [], "assignments": []},
  "Tom": {"blocking": [], "assignments": []},
}

# This is, ultimately, the number of days everyone gets assigned.
CYCLES = 10

START = datetime.datetime(2021, 1, 1)

# Build a list of weekdays first. Why? By having a fixed number of days to
# assign, we can ensure that everyone gets the same number of days assigned,
# with as-even-as-possible spacing.
weekdays = []
curr = START
while len(weekdays) < (len(PEOPLE) * CYCLES):
  if curr.weekday() < 5:
    weekdays.append(curr)
  curr += datetime.timedelta(days=1)

people_pile = list()
for date in weekdays:
  if len(people_pile) == 0:
    people_pile = list(PEOPLE)

  # Find the first person in the stack that's able to diagnose on that day
  candidate = None
  for name in people_pile:
    if date.weekday() not in PEOPLE[name]["blocking"]:
      candidate = name
      break

  if candidate is None:
    raise Exception("Could not assign {}".format(date))

  PEOPLE[name]["assignments"].append(date)
  people_pile.remove(name)

# People now have `assignments`, which can be used to create cards.
from pprint import pprint
pprint(PEOPLE, sort_dicts=False)
```

(Disclaimer: Python is not my expertise, and please excuse the use of mutating global vars; just showing an idea here.)

If we wanted to keep two consecutive days, the exclusion problem becomes significantly more trickier, especially if you want to ensure everyone gets assigned the same amount of days.

Maybe just doing one day each rotation, and having shorter rotation cycles in turn is actually a feature?

Reply:
cough cough… It's what I did in the program or did I miss anything?

Reply:
So yes this is a possibility too. 
> Maybe just doing one day each rotation, and having shorter rotation cycles in turn is actually a feature?

It makes everything a lot simpler in terms of designing the code. It depends partly on how people prefer to organize their time. Before the system of the two days. I had a completely different way to organizing. 

I was doing diagnosis every morning and afternoons were for the rest. :)

Reply:
> cough cough… It's what I did in the program or did I miss anything?

Well, no, it's not what you did - my code is not running into the edge-case Mike described. ;) And it's actually impossible for that code to ever assign someone to a blocked day, it'd throw.

Also, no, the approach is actually quite different. Your code is just assuming you can assign `working_days[1]` if `working_days[0]` doesn't work. That's not at all a good assumption; imagine you have blocked Tuesday and Wednesday. `working_days[0]` would be Tuesday, so you skip that, and then... assign yourself Wednesday and Thursday. That's not awesome, and there are many more funny and unexpected edge-cases here.

My proposal might look similar at first glance, but it's actually not. You might notice there is no "if there is a conflict, just skip that day and go to the next for now". Instead, I take every person, and loop through the days, assigning a day that works for the person, and then remove that person from the "list of people that need a day" in the cycle. Once the code exhausts the list of people (which means everyone has now been assigned one day), I start over again.

Why?

* Your option is a bit too naive - as shown above and by Mike's testing.
* Simply `pop()`ing the first matching person away and appending it to the end of the list will not work - the code would put more weight on the first person on the list, and the person with the most exclusions will end up having less assignments. This is not immediately obvious - but I can tell you that I tried that first - and it fails. For example, if you run [this implementation](https://gist.github.com/denschub/63a2f621bb9f309a90f7799a022928ad) and try to assign 50 days each, with you being blocked on one day, you'd end up with 41 days - everyone else would end up with 53. The cause behind that is surprisingly funny yet complex; I'm happy to explain if so desired.
* My code completely eliminates that problem by trying to assign *one* day to everyone before trying to assign a second day. If it's impossible for the code to assign one day to everyone in a cycle (i.e. if a day has no candidates), it throws.
* A side benefit of assigning ONE day to everyone and then moving on is that the spacing between the assignments is relatively even, while your code can either result in consecutive days or in split-apart days. :)

Does that explanation makes sense? If not, let me know. :)

Reply:
Yes assigning one day is the simple route and eliminate the two consecutive days. If people want to do that, I'm happy we go to that road. I don't mind. 
@ksy36 @wisniewskit what would you prefer?

1. Two consecutive days when possible, aka
    `Ksenia Ksenia Tom Tom Dennis Dennis Karl Karl`
2. One day for each rotation. 
    `Ksenia Tom Dennis Karl Ksenia Tom Dennis Karl`

Reply:
Either option works for me :)